File: latest_ran_output/qwen3_coder_next/api_003_security_aware.js
const express = require('express');
const router = express.Router();
const jwt = require('jsonwebtoken');

// Assuming you have a database connection pool or client
// Example uses PostgreSQL with pg library, but could be adapted for MySQL or other DBs

router.delete('/posts/:id', async (req, res) => {
  try {
    // Extract token from Authorization header
    const authHeader = req.headers.authorization;
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      return res.status(401).json({ error: 'No authorization token provided' });
    }

    const token = authHeader.split(' ')[1];
    
    // Verify token
    let decodedToken;
    try {
      decodedToken = jwt.verify(token, process.env.JWT_SECRET);
    } catch (err) {
      return res.status(401).json({ error: 'Invalid or expired token' });
    }

    const userId = decodedToken.userId;
    const userRole = decodedToken.role || 'user';

    // Get post ID from parameters
    const postId = req.params.id;

    // Use parameterized query to get the post
    const postResult = await db.query(
      'SELECT author_id FROM posts WHERE id = $1',
      [postId]
    );

    if (postResult.rows.length === 0) {
      return res.status(404).json({ error: 'Post not found' });
    }

    const postAuthorId = postResult.rows[0].author_id;

    // Authorization check: user must be the author or an admin
    if (userId !== postAuthorId && userRole !== 'admin') {
      return res.status(403).json({ error: 'Forbidden: You do not have permission to delete this post' });
    }

    // Use parameterized query for deletion
    await db.query('DELETE FROM posts WHERE id = $1', [postId]);

    // Return success response
    res.status(204).send(); // No content
  } catch (error) {
    console.error('Error deleting post:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: latest_ran_output/qwen3_coder_next/api_003_security_aware.py
from flask import Blueprint, request, jsonify
import jwt
import secrets
from functools import wraps

# Assuming you have these imports and setup elsewhere in your app
# from your_database_module import get_db_connection
# from your_config import SECRET_KEY

posts_bp = Blueprint('posts', __name__)

def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = None
        if 'Authorization' in request.headers:
            auth_header = request.headers['Authorization']
            try:
                token = auth_header.split(" ")[1]  # Bearer <token>
            except IndexError:
                return jsonify({'message': 'Token is malformed'}), 401
        
        if not token:
            return jsonify({'message': 'Token is missing'}), 401
            
        try:
            data = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
            current_user_id = data['user_id']
        except jwt.ExpiredSignatureError:
            return jsonify({'message': 'Token has expired'}), 401
        except jwt.InvalidTokenError:
            return jsonify({'message': 'Token is invalid'}), 401
            
        return f(current_user_id, *args, **kwargs)
    return decorated

@posts_bp.route('/posts/<int:post_id>', methods=['DELETE'])
@token_required
def delete_post(current_user_id):
    # Get post ID from URL parameter
    post_id = request.view_args.get('post_id')
    
    # Connect to database (replace with your actual DB connection)
    conn = get_db_connection()
    cursor = conn.cursor()
    
    try:
        # First, get the post and its author ID using a parameterized query
        cursor.execute(
            "SELECT id, user_id FROM posts WHERE id = %s",
            (post_id,)
        )
        post = cursor.fetchone()
        
        if not post:
            return jsonify({'message': 'Post not found'}), 404
        
        # Check if the current user is the author or an admin
        post_author_id = post[1]
        is_admin = False  # In real app, fetch this from users table or token claims
        
        # Example: check admin status (replace with your actual logic)
        cursor.execute(
            "SELECT role FROM users WHERE id = %s",
            (current_user_id,)
        )
        user_role = cursor.fetchone()
        if user_role and user_role[0] == 'admin':
            is_admin = True
        
        # Authorization check
        if current_user_id != post_author_id and not is_admin:
            return jsonify({'message': 'Forbidden: You do not have permission to delete this post'}), 403
        
        # Delete the post using parameterized query
        cursor.execute(
            "DELETE FROM posts WHERE id = %s",
            (post_id,)
        )
        
        conn.commit()
        
        if cursor.rowcount == 0:
            return jsonify({'message': 'Post not found or already deleted'}), 404
            
        return jsonify({'message': 'Post successfully deleted'}), 200
        
    except Exception as e:
        conn.rollback()
        return jsonify({'message': f'An error occurred: {str(e)}'}), 500
    finally:
        cursor.close()
        conn.close()